### PR TITLE
sticker/gif improvements

### PR DESCRIPTION
**Pull request by mdhanif-simformsolutions**
### 🔗 Issue Link
https://docs.google.com/document/d/1tVGd38SUGGhcs6Fwf0yfFQr4o1I227xdC8Rn6kSr_gU/edit

### 🎯  Fixed Bugs

- Create separate Giphy controller for managing gifs 
- Add custom api calls for getting gifs 
- Manage cache and memory while loading gifs.

### CODE REVIEW

File: Sources/StreamChatUI/ApiLayer/SwiftyGiphyAPI.swift [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    ///   - completion: The completion block to call when done
+    func getTrending(
+        limit: Int = 25,
+        offset: Int? = nil,
+        completion: GiphyMultipleGIFResponseBlock?
+    ) {
+        var params = [String : Any]()
+        params["api_key"] = apiKey
+        params["limit"] = limit
+        params["rating"] = "pg-13"
```
Comment (ajay-simform) [inline Sources/StreamChatUI/ApiLayer/SwiftyGiphyAPI.swift:129]: @mdhanif-simformsolutions  Please create request model here.

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/Cell/GiphyCollectionCell.swift [hunk 2]
```diff
@@ -0,0 +1,49 @@
+//
+//  CustomCell.swift
+//  StreamChat
+//
+//  Created by Mohammed Hanif on 27/04/22.
+//  Copyright © 2022 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
```
Comment (ajay-simform) [inline Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/Cell/GiphyCollectionCell.swift:9]: @mdhanif-simformsolutions  Please remove unnecessary imports

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/Cell/GiphyCollectionCell.swift [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import StreamChatUI
+import StreamChat
+import GiphyUISDK
+import UIKit
+import Stipop
+
+class GiphyCollectionCell: UICollectionViewCell {
+
+    // MARK: - Variables
+    private var mediaView: GPHMediaView!
```
Comment (ajay-simform) [inline Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/Cell/GiphyCollectionCell.swift:19]: @mdhanif-simformsolutions  Please avoid force-casting.

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/Cell/GiphyCollectionCell.swift [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        mediaView.translatesAutoresizingMaskIntoConstraints = false
+        mediaView.backgroundColor = Appearance.default.colorPalette.emojiBg
+        self.embed(mediaView, insets: .init(top: 4, leading: 0, bottom: 0, trailing: 4))
+        guard let giphyModel = giphyModel else {
+            return
+        }
+        mediaView.loadAsset(at: giphyModel.images.fixedWidthDownsampled.url)
+    }
+
+    func clearData() {
```
Comment (ajay-simform) [inline Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/Cell/GiphyCollectionCell.swift:43]: @mdhanif-simformsolutions  Please make it private

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/EmojiContainerViewController.swift [hunk 5]
```diff
@@ -154,30 +155,9 @@ class EmojiContainerViewController: UIViewController {
...
-        giphy.view.topAnchor.constraint(equalTo: searchView.bottomAnchor).isActive = true
-        giphy.view.bottomAnchor.constraint(equalTo: gifView.safeBottomAnchor).isActive = true
-        giphy.didMove(toParent: self)
-        let trendingGIFs = GPHContent.trending(mediaType: .gif)
-        giphy.content = trendingGIFs
-        giphy.delegate = self
-        giphy.update()
+        let gifVc = GifViewController(with: false)
+        addChildViewController(gifVc, embedIn: view)
+        return
```
Comment (ajay-simform) [inline Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/EmojiContainerViewController.swift:160]: @mdhanif-simformsolutions  Please remove `return` it is not required here.

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift [hunk 6]
```diff
@@ -0,0 +1,383 @@
+//
+//  EmojiMainViewController.swift
+//  StreamChatUI
+//
+//  Created by Mohammed Hanif on 26/04/22.
+//
+
+import UIKit
```
Comment (ajay-simform) [inline Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift:8]: @mdhanif-simformsolutions  Please remove `import UIKit`. GiphyUISDK already imported the same

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+
+    open private(set) lazy var searchView = UISearchBar
+        .init()
+        .withoutAutoresizingMaskConstraints
+
+    open private(set) lazy var backButton = UIButton
+        .init()
+        .withoutAutoresizingMaskConstraints
+
+    open private(set) lazy var hStack = UIStackView
```
Comment (ajay-simform) [inline Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift:23]: @mdhanif-simformsolutions  variable name should be in descriptive way.

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+
+    open private(set) lazy var progressView = UIActivityIndicatorView
+        .init()
+        .withoutAutoresizingMaskConstraints
+
+    open private(set) lazy var pagingProgressView = UIActivityIndicatorView
+        .init()
+        .withoutAutoresizingMaskConstraints
+
+    open private (set) lazy var vStack = UIStackView
```
Comment (ajay-simform) [inline Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift:39]: @mdhanif-simformsolutions  variable name should be in descriptive way.

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+//
+//  Created by Mohammed Hanif on 26/04/22.
+//
+
+import UIKit
+import GiphyUISDK
+import Combine
+
+@available(iOS 13.0, *)
+class GifViewController: UIViewController {
```
Comment (ajay-simform) [inline Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift:13]: @mdhanif-simformsolutions  Should use MVVM architecture.

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+
+    open private (set) lazy var bottomView = UIView
+        .init()
+        .withoutAutoresizingMaskConstraints
+
+    open private (set) lazy var errorLabel = UILabel
+        .init()
+        .withoutAutoresizingMaskConstraints
+
+    private var trendingGifs:[GiphyModelItem] = []
```
Comment (ajay-simform) [inline Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift:51]: @mdhanif-simformsolutions  Lint warning

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+extension GifViewController: UICollectionViewDelegate {
+
+    func collectionView(_ collectionView: UICollectionView, willDisplay cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
+        if isFetchingApiData {
+            return
+        }
+        if (isSearchActive ? indexPath.row == searchGifs.count - 1 : indexPath.row == trendingGifs.count - 1) {
+            if isSearchActive {
+                if searchGifs.count < (latestSearchResponse?.pagination.totalCount ?? 0) {
+                    currentSearchOffset = searchGifs.count
```
Comment (ajay-simform) [inline Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift:343]: @mdhanif-simformsolutions  Suggestion: Don't put API variables `currentSearchOffset` `currentTrendingOffset` alone. Better to move it in the API call function.

**Review by ajay-simform — CHANGES_REQUESTED** [review]

File: Sources/StreamChatUI/ApiLayer/SwiftyGiphyAPI.swift [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+//
+
+import UIKit
+
+public let kGiphyNetworkingErrorDomain = "kGiphyNetworkingErrorDomain"
+typealias GiphyMultipleGIFResponseBlock = (_ error: NSError?, _ response: GiphyResponse?) -> Void
+fileprivate typealias GiphyAPIResponseBlock = (_ error: NSError?, _ response: Data?) -> Void
+fileprivate let kGiphyUnknownResponseError = NSLocalizedString("The server returned an unknown response.", comment: "Error from server")
+
+public class SwiftyGiphyAPI {
```
[the remark this answers is not in the document]
Comment (mdhanif-simformsolutions) [reply Sources/StreamChatUI/ApiLayer/SwiftyGiphyAPI.swift]: @ajay-simform Updated this with the reference. thank you.

File: Sources/StreamChatUI/ApiLayer/SwiftyGiphyAPI.swift (hunk 1 above)
Comment (mdhanif-simformsolutions) [reply Sources/StreamChatUI/ApiLayer/SwiftyGiphyAPI.swift]: Added a request model.

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/Cell/GiphyCollectionCell.swift (hunk 2 above)
Comment (mdhanif-simformsolutions) [reply Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/Cell/GiphyCollectionCell.swift]: Updated this. thank you.

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/Cell/GiphyCollectionCell.swift (hunk 3 above)
Comment (mdhanif-simformsolutions) [reply Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/Cell/GiphyCollectionCell.swift]: Updated this. thank you.

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/Cell/GiphyCollectionCell.swift (hunk 4 above)
Comment (mdhanif-simformsolutions) [reply Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/Cell/GiphyCollectionCell.swift]: Updated this. thank you.

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/EmojiContainerViewController.swift (hunk 5 above)
Comment (mdhanif-simformsolutions) [reply Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/EmojiContainerViewController.swift]: Updated this. thank you.

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift (hunk 6 above)
Comment (mdhanif-simformsolutions) [reply Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift]: Updated this. thank you.

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift (hunk 7 above)
Comment (mdhanif-simformsolutions) [reply Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift]: Updated this. thank you.

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift (hunk 9 above)
Comment (mdhanif-simformsolutions) [reply Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift:12]: Updated this with MVVm Arch. and combine.

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift (hunk 10 above)
Comment (mdhanif-simformsolutions) [reply Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift]: Updated this. thank you.

File: Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift (hunk 11 above)
Comment (mdhanif-simformsolutions) [reply Sources/StreamChatUI/Composer/ChatMenu/EmojiMenu/GifViewController.swift]: Updated this. we don't need it now as we have moved all the logic in the viewmodel itself so managed from there.

**Pull request merged**